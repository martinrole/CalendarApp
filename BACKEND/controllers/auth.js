const { response } = require('express')
const Usuario = require('../models/Usuario')
const bcrypt = require('bcryptjs')
const { generarJWT } = require('../helpers/jwt')


const crearUsuario = async (req, res = response ) => {
    
    const { nombre, correo, contrasena } = req.body

    try {

        let usuario = await Usuario.findOne({ correo: correo })
        
        if ( usuario ) {
            return res.status(400).json({
                ok: false,
                msg: 'correo ya fue registrado anteriormente'
            })
        }

        //Guarda usuario:
        usuario = new Usuario(req.body)
        const salt = bcrypt.genSaltSync(12)
        usuario.contrasena = bcrypt.hashSync(contrasena, salt)
        await usuario.save()

        //Generar JWT:
        const token = await generarJWT( usuario.id, usuario.nombre )
    
        //Respuesta de petici칩n:
        res.status(201).json({
            ok: true,
            msg: 'user created',
            uid: usuario.id,
            nombre: usuario.nombre,
            token
        })
        
    } catch (error) {
        console.log('crearUsuario: ', error);
        res.status(500).json({
            ok: false,
            msg: 'Hubo un error creando el usuario'
        })
    }

}


const loginUsuario = async (req, res = response) => {
     
    const { correo, contrasena } = req.body

    try {
        
        //Valida que exista un correo
        const usuario = await Usuario.findOne({ correo: correo })
        
        if ( !usuario ) {
            return res.status(400).json({
                ok: false,
                msg: 'Usuario y/o contrase침a no son correctos'
            })
        }

        //Valida las contrase침as:
        const validPass = bcrypt.compareSync(contrasena, usuario.contrasena)

        if (!validPass){
            return res.status(400).json({
                ok: false,
                msg: 'Usuario y/o contrase침a no son correctos 2'
            })
        }

        //Generar JWT:
        const token = await generarJWT( usuario.id, usuario.nombre )

        //Generar JSON respuesta correcta:
        return res.status(201).json({
            ok: true,
            msg: 'login completed',
            uid: usuario.id,
            nombre: usuario.nombre,
            token
        })

    } catch (error) {
        console.log('loginUsuario: ', error);
        res.status(500).json({
            ok: false,
            msg: 'Hubo un error al realizar login'
        })
    }
    
}


const revalidarToken =  async (req, res = response) => {

    //Viene ya ajustado del middelware validarJWT aplicado a la ruta:
    const { uid, nombre } = req

    //Genera nuevo JWT:
    const token = await generarJWT( uid, nombre )

    res.json({
        ok: true,
        msg: 'update token',
        token,
        uid,
        nombre
    })
}


module.exports = {
    crearUsuario,
    loginUsuario,
    revalidarToken
}