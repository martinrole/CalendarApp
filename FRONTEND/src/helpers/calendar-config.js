import moment from "moment"

//Pone los titulos del calendario en español
export const calendarConfig = {
    allDay: 'Todo el día',
    previous: 'Atrás',
    next: 'Siguiente',
    today: 'Hoy',
    month: 'Mes',
    week: 'Semana',
    day: 'Día',
    agenda: 'Agenda',
    date: 'Fecha',
    time: 'Hora',
    event: 'Evento',
    noEventsInRange: 'No hay eventos en este rango',
    showMore: total => `+ Ver más (${total})`
};

//Video 379: Convierte la fecha String del Backend en una válida para el Calendario
export const adjustEventsDate = ( eventos = [] ) => { 
   return eventos.map( evento => ({
            ...evento, 
            start: moment(evento.start).toDate(),
            end: moment(evento.end).toDate()
        })
    )
}
