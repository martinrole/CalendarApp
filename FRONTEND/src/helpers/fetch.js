const baseUrl = process.env.REACT_APP_API_URL

//Estas funciones se ejecutas dependiendo si es una petición privada o no

//Autenticación sin token
export const authFetch = async ( url, data, method ) => {
    const urlReq = `${baseUrl}/${url}`

    if ( method === 'GET') {
        return await fetch(urlReq)
    }

    const params = {
        method,
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify( data )
    }

    try {
        const respuesta = await fetch( urlReq, params)
        return await respuesta.json()

    } catch (error) {
        console.log('error en authFetch: ', error);
        return null
    }
}


//Autenticación con token: Comprueba que el usuario esté logueado y el token vigente:
export const authFetchToken = async ( url, data, method ) => {
    const urlReq = `${baseUrl}/${url}`
    const token = localStorage.getItem('token') || ''
    let params

    if (method === 'GET') {
        params = {
            method,
            headers: {
                'x-token': token
            }
        }
    } else {
        params = {
            method,
            headers: {
                'Content-Type': 'application/json',
                'x-token': token
            },
            body: JSON.stringify( data )
        }
    }


    try {
        const respuesta = await fetch( urlReq, params)
        return await respuesta.json()

    } catch (error) {
        console.log('error en authFetchToken: ', error);
        return null
    }
}