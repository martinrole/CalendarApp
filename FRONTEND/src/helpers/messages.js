import Swal from 'sweetalert2'

export const loadingMessage = (state) => {

    if (state) {
        Swal.fire({
            html: 'Cargando',
            didOpen: () => {
                Swal.showLoading()
            }
        })
    } else {
        Swal.close()
    }
}

export const showMessage =(icon,title, text) => {
    Swal.fire({
        icon,
        title,
        text
    })
}