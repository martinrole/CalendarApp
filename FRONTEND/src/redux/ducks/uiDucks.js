

//1- initialState:
export const initialState = {
    modalOpen: false
}

//2 - Types:
export const types = {
    uiOpenModal: '[uiDucks] Open Modal',
    uiCloseModal: '[uiDucks] Close Modal'
}


//3 - Reducer:
export const uiReducer = (state = initialState, action ) => {
    switch (action.type) {
        case types.uiOpenModal:
            return { ...state, modalOpen: true }
        case types.uiCloseModal:
                return { ...state, modalOpen: false }
    
        default:
            return state;
    }
}

//4 - Actions:
export const openModalAction = () => ({
    type: types.uiOpenModal
})

export const closeModalAction = () => ({
    type: types.uiCloseModal
})