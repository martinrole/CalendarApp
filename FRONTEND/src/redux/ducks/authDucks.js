import { authFetch, authFetchToken } from "../../helpers/fetch";
import { showMessage } from "../../helpers/messages";


//1- initialState:
const initialState = {
    checking: true,
}

//2 - Types:
export const types = {
    checkingFinish: '[authDucks] Checking Finish',
    startLogin: '[authDucks] Start Login',
    login: '[authDucks] Login',
    startRegister: '[authDucks] Start Register',
    startTokenRenew: '[authDucks] Start Token renew',
    logout: '[authDucks] Logout'
}


//3 - Reducer:
export const authReducer = (state = initialState, action ) => {
    switch (action.type) {
        case types.login:
            return { 
                ...state, 
                ...action.payload,
                checking: false
            }
        case types.checkingFinish:
            return {
                ...state,
                checking: false
            }
        case types.logout:
            return {
                checking: false
            }
    
        default:
            return state;
    }
}

//4 - Actions:

//Primera forma de hacer:
// export const loginAction = ( correo, contrasena ) => {
//     return async ( dispatch ) => {
//         console.log('haber', correo, contrasena);
//     }
// }

//Segunda forma de hacerlo:
export const loginAction = ( correo, contrasena ) => async ( dispath, getState ) => {
    const data = { correo, contrasena }
    const { uid, nombre, token, ok, msg } = await authFetch('auth', data, 'POST')
    
    if ( ok ) {
        localStorage.setItem('token',token)
        localStorage.setItem('token-init-date', new Date().getTime())

        dispath({
            type: types.login,
            payload: {
                uid,
                nombre
            }
        })

    } else {
        showMessage('error', null, msg )
    }

}

export const registerAction = ( correo, name, contrasena ) => async ( dispath, getState) => {

    const data = { correo, nombre: name, contrasena }
    const { uid, nombre, token, ok, msg } = await authFetch('auth/create', data, 'POST') 

    if ( ok ) {
        localStorage.setItem('token',token)
        localStorage.setItem('token-init-date', new Date().getTime())

        dispath({
            type: types.login,
            payload: {
                uid,
                nombre
            }
        })

        showMessage('success', 'Bienvenido','Registro exitoso')

    } else {
        showMessage('error', null, msg )
    }
}

export const authCheckAction = () => async ( dispath, getState) => {

    const res = await authFetchToken('auth/updateToken', null , 'GET')
    
    if ( res.ok ) {
        localStorage.setItem('token',res.token)
        localStorage.setItem('token-init-date', new Date().getTime())

        dispath({
            type: types.login,
            payload: {
                uid: res.uid,
                nombre: res.nombre
            }
        })

    } else {
        dispath({
            type: types.checkingFinish
        })
    }
}

export const logoutAction = () =>  (dispath, getState) => {
    dispath({
        type: types.logout
    })
    
    localStorage.clear()
}