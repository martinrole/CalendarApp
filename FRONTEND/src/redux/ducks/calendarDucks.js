import { adjustEventsDate } from "../../helpers/calendar-config"
import { authFetchToken } from "../../helpers/fetch"
import { showMessage } from "../../helpers/messages"




// Lo use para antes de conectar con la base de datos:
// const initialState = {
//     events: [
//         {
//             id: 'sdsd&hnsd56gGRf123bSD',
//             title: 'PruebaReducer',
//             start: moment().toDate(),     //Sinónimo de new Date() de JavaScript
//             end: moment().add(1, 'hours').toDate(),
//             allDay: false,
//             notes: 'acá van las notas',
//             user: {
//                 _id: '12dsfv324D',
//                 nombre: 'Martin User'
//             }
//         }
//     ],
//     activeEvent: null
// }

//1- initialState:
const initialState = {
    events: [],
    activeEvent: null
}

//2 - Types:
const types = {
    eventNew: '[CalendarDucks] Add Event',
    eventList: '[CalendarDucks] List Event',
    eventActive: '[CalendarDucks] Active Event',
    eventClearActive: '[CalendarDucks] Clear Active Event',
    eventUpdate: '[CalendarDucks] Updated Event',
    eventDelete: '[CalendarDucks] Delete Event',
    clearCalendar: '[CalendarDucks] Clear All Calendar'
}

//3 - Reducer:
export const calendarReducer = (state = initialState, action ) => {
    switch (action.type) {
        case types.eventActive:
            return { 
                ...state, 
                activeEvent: action.payload 
            }
        case types.eventClearActive:
            return { 
                ...state, 
                activeEvent: null 
            }
        case types.eventNew:
            return { 
                ...state, 
                events: [...state.events, action.payload]
            }
        case types.eventList:
            return {
                ...state, 
                events: [ ...action.payload ] 
            }
        case types.eventUpdate:
            return {
                ...state, 
                events: state.events.map( event => event.id === action.payload.id ? action.payload : event )
            }
        case types.eventDelete:
            return {
                ...state, 
                events: state.events.filter( event => event.id !== state.activeEvent.id ),
                activeEvent: null
            }
        case types.clearCalendar:
            return {
                ...initialState
            }
        default:
            return state;
    }
}

//4 - Actions:
export const eventNewAction = ( event ) => async ( dispatch, getState ) => {
    
    const { uid, nombre } = getState().auth

    try {   
        const res = await authFetchToken('events/create', event, 'POST')

        const evento = {
            id: res.evento.id,
            ...event,
            usuario: {
                _id: uid,
                nombre
            }
        }
    
        if ( res.ok ) {
            dispatch({
                type: types.eventNew,
                payload: evento
            })
        }

    } catch (error) {
        console.log('error en eventNewAction', error);
        
    }
}


export const eventLoadAction = () => async ( dispatch, getState ) => {

    try {
        const { ok, eventos } = await authFetchToken('events/', null, 'GET')
        
        if (ok) {        
            const events = adjustEventsDate( eventos )
            dispatch({
                type: types.eventList,
                payload: events
            })
        }

    } catch (error) {
        console.log('error en eventLoadAction', error);
        
    }
}

export const eventUpdateAction = ( event ) => async (dispatch, getState ) => {

    try {
        const { ok, msg } = await authFetchToken(`events/${event.id}`, event, 'PUT')

        if ( ok ) {
            dispatch({
                type: types.eventUpdate,
                payload: event
            })
        } else {
            //Solo se pueden editar los eventos propios del user
            showMessage('error', null, msg)
        }

    } catch (error) {
        console.log('eventUpdateAction', error);
        
    }
}

export const eventDeleteAction = () => async (dispatch, getState ) => {

    try {
        const { id } = getState().calendar.activeEvent
        const { ok, msg } = await authFetchToken(`events/${id}`, {}, 'DELETE')

        if ( ok ) {
            dispatch({
                type: types.eventDelete
            })
        } else {
            showMessage('error', null, msg)
        }

    } catch (error) {
        console.log('error eventDeleteAction', error);   
    }
    
}

export const clearCalendarAction = () => (dispatch, getState ) => {
    console.log('clearCalendarAction: ', clearCalendarAction);
    dispatch({
        type: types.clearCalendar
    })
}

export const eventActiveAction = ( event ) => ({
    type: types.eventActive,
    payload: event
})

export const eventClearActiveAction = () => ({
    type: types.eventClearActive
})
