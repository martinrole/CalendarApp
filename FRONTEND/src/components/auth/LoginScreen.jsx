import { useForm } from '../../hooks/useForm';
import { useDispatch } from 'react-redux'
import './login.css';
import { loginAction, registerAction } from '../../redux/ducks/authDucks';
import { showMessage } from '../../helpers/messages';

export const LoginScreen = () => {

    //useForm:
    const initialForm = {
        loginCorreo: 'martinrole@hotmail.com',
        loginContrasena: 'mar123',
        rNombre: 'Tobby',
        rCorreo: 'tobbias23@hotmail.com',
        rContrasena: 'soyTobby1a',
        rContrasena2: 'soyTobby1a',
    };
    const [formValues, handleInputChange] = useForm(initialForm);
    const { loginCorreo, loginContrasena, rNombre, rCorreo, rContrasena, rContrasena2 } = formValues

    //dispatch:
    const dispatch = useDispatch()

    const handleLogin = (e) => {
        e.preventDefault()

        dispatch(loginAction(loginCorreo, loginContrasena))
    }

    const handleRegister = (e) => {
        e.preventDefault()

        if (rContrasena !== rContrasena2) {
            return showMessage('error', null, 'Contrase単as no son iguales')
        }

        dispatch(registerAction(rCorreo, rNombre, rContrasena))
    }

    return (
        <div className="container login-container">
            <div className="row">
                <div className="col-md-6 login-form-1">
                    <h3>Ingreso</h3>
                    <form onSubmit={handleLogin}>
                        <div className="form-group">
                            <input
                                type="text"
                                name="loginCorreo"
                                className="form-control"
                                placeholder="Correo"
                                value={loginCorreo}
                                onChange={handleInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                name="loginContrasena"
                                className="form-control"
                                placeholder="Contrase単a"
                                value={loginContrasena}
                                onChange={handleInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="submit"
                                className="btnSubmit"
                                value="Ingresar"
                            />
                        </div>
                    </form>
                </div>

                <div className="col-md-6 login-form-2">
                    <h3>Registro</h3>
                    <form onSubmit={handleRegister}>
                        <div className="form-group">
                            <input
                                type="text"
                                name="rNombre"
                                className="form-control"
                                placeholder="Nombre"
                                value={rNombre}
                                onChange={handleInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="email"
                                name="rCorreo"
                                className="form-control"
                                placeholder="Correo"
                                value={rCorreo}
                                onChange={handleInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                name="rContrasena"
                                className="form-control"
                                placeholder="Contrase単a"
                                value={rContrasena}
                                onChange={handleInputChange}
                            />
                        </div>

                        <div className="form-group">
                            <input
                                type="password"
                                name="rContrasena2"
                                className="form-control"
                                placeholder="Repita la contrase単a"
                                value={rContrasena2}
                                onChange={handleInputChange}
                            />
                        </div>

                        <div className="form-group">
                            <input
                                type="submit"
                                className="btnSubmit"
                                value="Crear cuenta" />
                        </div>
                    </form>
                </div>
            </div>
        </div>
    )
}