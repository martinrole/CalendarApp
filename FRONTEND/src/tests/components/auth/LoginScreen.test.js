import React from 'react'
import { mount } from 'enzyme'
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store' //ES6 modules
import thunk from 'redux-thunk';
import { LoginScreen } from '../../../components/auth/LoginScreen';
import { loginAction, registerAction } from '../../../redux/ducks/authDucks';
import { showMessage } from '../../../helpers/messages';

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initialState = {}

const store = mockStore(initialState)
store.dispatch = jest.fn()

const wrapper = mount(
    <Provider store={store}>
        <LoginScreen />
    </Provider>
)

jest.mock('../../../redux/ducks/authDucks', () => ({
    loginAction: jest.fn(),
    registerAction: jest.fn()
}))

jest.mock('../../../helpers/messages', () => ({
    showMessage: jest.fn()
}))


describe('Prueba en <LoginScreen />', () => {
    
    //Siempre que se utilicen mocks es recomendale hacer un beforeEach:
    beforeEach(() => {
        jest.clearAllMocks()
    })

    test('Debe mostrar el componente correctamente', () => {
        expect( wrapper ).toMatchSnapshot()
    });

    test('Debe llamar el dispatch del Login', () => {

        wrapper.find('input[name="loginCorreo"]').simulate('change', {
            target: {
                name: 'loginCorreo',
                value: 'correo_test@testing.com'
            }
        })

        wrapper.find('input[name="loginContrasena"]').simulate('change', {
            target: {
                name: 'loginContrasena',
                value: 'password_1234'
            }
        })

        //Explicaci칩n video 401 por si algo:
        wrapper.find('form').at(0).prop('onSubmit')({
            preventDefault(){}
        })

        expect( loginAction ).toHaveBeenCalledWith('correo_test@testing.com','password_1234')

    });

    //Explicaci칩n video 402:
    test('No hay registro si las contrase침as son diferentes', () => {
        
        wrapper.find('input[name="rContrasena"]').simulate('change', {
            target: {
                name: 'rContrasena',
                value: 'password_123'
            }
        })

        wrapper.find('input[name="rContrasena2"]').simulate('change', {
            target: {
                name: 'rContrasena2',
                value: 'password_123456'
            }
        })

        //Segundo formulario, por eso pongo at(1)
        wrapper.find('form').at(1).prop('onSubmit')({
            preventDefault(){}
        })

        expect( showMessage ).toHaveBeenCalledWith("error", null, "Contrase침as no son iguales")
        expect( registerAction ).toHaveBeenCalledTimes(0)
        expect( registerAction ).not.toHaveBeenCalled() //Otra forma de hacerlo
    });

    test('Registrarse un usuario correctamente', () => {
        wrapper.find('input[name="rContrasena"]').simulate('change', {
            target: {
                name: 'rContrasena',
                value: 'password_123'
            }
        })

        wrapper.find('input[name="rContrasena2"]').simulate('change', {
            target: {
                name: 'rContrasena2',
                value: 'password_123'
            }
        })

        //Segundo formulario, por eso pongo at(1)
        wrapper.find('form').at(1).prop('onSubmit')({
            preventDefault(){}
        })

        expect( showMessage ).not.toHaveBeenCalled()
        expect( registerAction ).toHaveBeenCalledTimes(1)
        expect( registerAction ).toHaveBeenCalledWith('tobbias23@hotmail.com', 'Tobby', 'password_123')
    });
})