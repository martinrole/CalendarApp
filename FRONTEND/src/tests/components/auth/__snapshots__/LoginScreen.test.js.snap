// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en <LoginScreen /> Debe mostrar el componente correctamente 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Ingreso
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginCorreo"
            onChange={[Function]}
            placeholder="Correo"
            type="text"
            value="martinrole@hotmail.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginContrasena"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="mar123"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Ingresar"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rNombre"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value="Tobby"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rCorreo"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value="tobbias23@hotmail.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rContrasena"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="soyTobby1a"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rContrasena2"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value="soyTobby1a"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
